crates/matrix-sdk-common/src/linked_chunk/as_vector.rs

```diff
@@ -137,6 +138,17 @@ impl<Item, Acc: UpdatesAccumulator<Item>> UpdateToVectorDiff<Item, Acc> {
         Self { chunks: initial_chunk_lengths, _phantom: std::marker::PhantomData }
     }
 
+    /// Construct [`UpdateToVectorDiff`], based on a linked chunk's full
+    /// metadata, used to set up its own internal state.
+    ///
+    /// See [`Self::map`] to learn more about the algorithm.
+    pub fn from_metadata(metas: Vec<ChunkMetadata>) -> Self {
+        let initial_chunk_lengths =
+            metas.into_iter().map(|meta| (meta.identifier, meta.num_items)).collect();
```

Comment:
Not a blocker!

We are doing a double allocation here. One `collect` in `from_metadata`, and one `collect` in `LinkedChunk::order_tracker`. I propose the following:

```suggestion
    pub fn from_metadata<M>(metas: M) -> Self
        where M: Iterator<Item = ChunkMetadata>
    {
        let initial_chunk_lengths =
            metas.map(|meta| (meta.identifier, meta.num_items)).collect();
```

and in `order_tracker`, we pass the iterator directly, without calling `collect`. I don't know how it will work with the `unwrap_or_else` though.